test/apps/iodemo/io_demo.cc

```diff
@@ -1135,6 +1136,21 @@ class DemoClient : public P2pDemoCommon {
         _prev_connect_time = curr_time;
     }
 
+    size_t get_server_index() {
+        size_t active_index, server_index;
+
+        do {
+            /* Pick a random connected server to which the client has credits
+             * to send (its conn's window is not full) */
+            active_index = IoDemoRandom::rand(size_t(0),
+                                              _active_servers.size() - 1);
+            server_index = _active_servers[active_index];
+        } while (get_num_uncompleted(server_index) > opts().conn_window_size);
```

Comment:
Seems like a hang if connection_window * N_servers < shared_window, i.e. all connections are full but shared window is not yet

Reply:
yes but, the change in the value passed to wait_for_responses should make progress and allow for completions to prevent the hang.

```wait_for_responses(std::min(opts().window_size - 1,opts().conn_window_size * (long)_active_servers.size()));```

Reply:
ic then `max_outstanding` should be
` std::min(opts().window_size, opts().conn_window_size * (long)_active_servers.size()) - 1` to have at least 1 free slot

maybe introduce index of first "full" server in `_active_server` arr to avoid random misses?
we could swap first "full" element with "free" server on completion, so here this index must be > 0 here and random range [0; full_index) could allow to get free server from first hit

Reply:
we need a vector of all servers which are available to send and select random among them
instead of doing a while loop like this
since if we have only few servers avail out of many this loop can take a long time

Reply:
we can "sort" existing `_active_servers` vector per completion and maintain 1 extar index instead of adding new vector like
struct {
std::vector<size_t> _all;
size_t                       _full_server_index; // 0 - all servers are full, _all.size() - all servers are able to send
} _active_servers_indexes;

Reply:
this approach is cheaper from performance perspective O(1) since maintaining another container adds overhead:
 - vector O(N) mem move per erase 
 - list O(1) but with extra allocation per insert and free per erase

Reply:
ok, makes sense. @yosefe ?

Reply:
+1 to use active_servers